Libraries/LibWeb/Encoding/TextEncoderStream.cpp

```diff
@@ -0,0 +1,219 @@
+    : Bindings::PlatformObject(realm)
+    , Streams::GenericTransformStreamMixin(transform)
+{
+}
+
+TextEncoderStream::~TextEncoderStream() = default;
+
+void TextEncoderStream::initialize(JS::Realm& realm)
+{
+    Base::initialize(realm);
+    WEB_SET_PROTOTYPE_FOR_INTERFACE(TextEncoderStream);
+}
+
+void TextEncoderStream::visit_edges(JS::Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    Streams::GenericTransformStreamMixin::visit_edges(visitor);
+}
+
+// https://encoding.spec.whatwg.org/#encode-and-enqueue-a-chunk
+WebIDL::ExceptionOr<void> TextEncoderStream::encode_and_enqueue_chunk(JS::Value chunk)
+{
+    // Spec Note: This is equivalent to the "convert a string into a scalar value string" algorithm from the Infra
+    //            Standard, but allows for surrogate pairs that are split between strings. [INFRA]
+
+    auto& realm = this->realm();
+    auto& vm = this->vm();
+
+    // 1. Let input be the result of converting chunk to a DOMString.
+    auto input = TRY(chunk.to_string(vm));
+
+    // 2. Convert input to an I/O queue of code units.
+    // Spec Note: DOMString, as well as an I/O queue of code units rather than scalar values, are used here so that a
+    //            surrogate pair that is split between chunks can be reassembled into the appropriate scalar value.
+    //            The behavior is otherwise identical to USVString. In particular, lone surrogates will be replaced
+    //            with U+FFFD.
+    auto code_points = input.code_points();
+    auto it = code_points.begin();
+
+    // 3. Let output be the I/O queue of bytes « end-of-queue ».
+    ByteBuffer output;
+
+    // 4. While true:
+    while (true) {
+        // 2. If item is end-of-queue, then:
+        // NOTE: This is done out-of-order so that we're not dereferencing a code point iterator that points to the end.
+        if (it.done()) {
+            // 1. Convert output into a byte sequence.
+            // Note: No-op.
+
+            // 2. If output is non-empty, then:
+            if (!output.is_empty()) {
+                // 1. Let chunk be a Uint8Array object wrapping an ArrayBuffer containing output.
+                auto array_buffer = JS::ArrayBuffer::create(realm, move(output));
+                auto array = JS::Uint8Array::create(realm, array_buffer->byte_length(), *array_buffer);
+
+                // 2. Enqueue chunk into encoder’s transform.
+                TRY(Streams::transform_stream_default_controller_enqueue(*m_transform->controller(), array));
+            }
+
+            // 3. Return.
+            return {};
+        }
+
+        // 1. Let item be the result of reading from input.
+        auto item = *it;
+
+        // 3. Let result be the result of executing the convert code unit to scalar value algorithm with encoder, item and input.
+        // https://encoding.spec.whatwg.org/#ref-for-convert-code-unit-to-scalar-value
+        auto result = [&] -> Optional<u32> {
```

Comment:
Can we make this its own function instead of inlining the whole thing as lambda?
The spec link should also be: https://encoding.spec.whatwg.org/#convert-code-unit-to-scalar-value
(you've linked the reference)